### PR TITLE
[SPARK-4962] [CORE] Put TaskScheduler.start back in SparkContext to shorten cluster resources occupation period

### PR DESCRIPTION
**Pull request by YanTangZhai**
When SparkContext object is instantiated, TaskScheduler is started and some resources are allocated from cluster. However, these
resources may be not used for the moment. For example, DAGScheduler.JobSubmitted is processing and so on. These resources are wasted in
this period. Thus, we want to put TaskScheduler.start back to shorten cluster resources occupation period specially for busy cluster.
TaskScheduler could be started just before running stages.
We could analyse and compare the resources occupation period before and after optimization.
TaskScheduler.start execution time: [time1__]
DAGScheduler.JobSubmitted (excluding HadoopRDD.getPartitions or TaskScheduler.start) execution time: [time2_]
HadoopRDD.getPartitions execution time: [time3___]
Stages execution time: [time4_____]
The cluster resources occupation period before optimization is [time2_][time3___][time4_____].
The cluster resources occupation period after optimization is....[time3___][time4_____].
In summary, the cluster resources
occupation period after optimization is less than before.
If HadoopRDD.getPartitions could be put forward (SPARK-4961), the period may be shorten more which is [time4_____].
The resources saving is important for busy cluster.

The main purpose of this PR is to decrease resources waste for busy cluster.
For example, a process initializes a SparkContext instance, reads a few files from HDFS or many records from PostgreSQL, and then calls RDD's collect operation to submit a job.
When SparkContext is initialized, an app is submitted to cluster and some resources are hold by this app. 
These resources are not used really until the job is submitted by RDD's action.
The resources in the period from initialization to actual use could be considered wasteful.
If app is submitted when SparkContext is initialized, all of resources needed by the app may be granted before running job. 
Then the job could runs efficiently without resource constraint.
On the contrary, if app is submitted when job is submitted, resources needed by the app may be granted at different times. Then the job may run not so efficiently since some resources are applying.
Thus I use a configuration parameter spark.scheduler.app.slowstart (default false) to let user make tradeoffs between economy and efficiency.
There are 9 kinds of master URL and 6 kinds of SchedulerBackend.
LocalBackend and SimrSchedulerBackend don't need to put starting back since there is no difference.
SparkClusterSchedulerBackend (yarn-standalone or yarn-cluster) does not put starting back since the app should be submitted in advance by SparkSubmit.
CoarseMesosSchedulerBackend and MesosSchedulerBackend could put starting back.
YarnClientSchedulerBackend (yarn-client) could put starting back.
This PR puts TaskScheduler.start back only for yarn-client mode in the early.

### CODE REVIEW

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 1]
```diff
@@ -55,13 +57,9 @@ private[spark] class Client(
...
   /**
-   * Submit an application running our ApplicationMaster to the ResourceManager.
-   *
-   * The stable Yarn API provides a convenience method (YarnClient#createApplication) for
-   * creating applications and setting up the application submission context. This was not
-   * available in the alpha API.
+   * Create an application running our ApplicationMaster to the ResourceManager.
    */
-  override def submitApplication(): ApplicationId = {
+  override def createApplication(): ApplicationId = {
```
Comment (YanTangZhai) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: SparkContext firstly gets applicationId from taskScheduler and uses it to initialize blockManager and eventLogger. And then dagScheduler runs job and submits resources requests to cluster master.
Getting applicationId and submitting resources requests to cluster master are split into two methods.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 2]
```diff
@@ -333,9 +333,15 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       new SparkException("DAGScheduler cannot be initialized due to %s".format(e.getMessage))
   }
 
-  // start TaskScheduler after taskScheduler sets DAGScheduler reference in DAGScheduler's
-  // constructor
-  taskScheduler.start()
+  if (conf.getBoolean("spark.scheduler.app.slowstart", false) && master == "yarn-client") {
```
Comment (YanTangZhai) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: I'm sorry that I couldn't get mesos apis to split getting applicationId and submitting resources requests to cluster master into two methods.

Comment (YanTangZhai) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: I'm sorry that I couldn't get mesos apis to split getting applicationId and submitting resources requests to cluster master into two methods.

**Comment by YanTangZhai** [comment]
@srowen I've updated this PR and resolved conflict. Please review again. Thanks.
I explain three points:
1. I am not sure the description makes a case that it's significant enough to bother...
Let me give two examples:
(1) When I entered ./bin/spark-sql in command line with yarn-client mode and these resources requests as follows 
spark.executor.instances 100
spark.executor.memory 4g
spark.executor.cores 1.
However, I didn't enter sql query string immediately. Because I was interrupted for example I was called to attend a important meeting or I go to fire fighting in our cluster. Even sometimes I forgot enter sql query string.
Then this application ran a night using 100 * 4g * 12h memory resources and 100 * 1 * 12h core resources. But it did nothing.
(2) When SparkContext with 100 spark.executor.instances、4g spark.executor.memory、1 spark.executor.cores was initialized and HadoopRDD scanned 11596 files taking 29.253s to compute splits. And then this job was submitted by DAGScheduler. The resources of 100 * 4g * 29s memory resources and 100 * 1 * 29s core resources were idle.
2. There are several new API methods and changes here.
SparkContext firstly gets applicationId from taskScheduler and uses it to initialize blockManager and eventLogger. And then dagScheduler runs job and submits resources requests to cluster master.
Getting applicationId and submitting resources requests to cluster master are split into two methods.
3. My overall impression is that this adds different code paths and behaviors in different modes for little gain.
I'm sorry that I couldn't get mesos apis to split getting applicationId and submitting resources requests to cluster master into two methods.
Thus slow start of application is currently only supported in YARN mode.

**Comment by srowen** [comment]
I don't think this is going to be merged. Do you mind closing this PR?

**Comment by andrewor14** [comment]
@YanTangZhai would you mind closing this?

**Pull request closed without merging**